### ISSUE TITLE
CustomNameExtractor issues with property override

### ISSUE THREAD

**Issue by asherber** [d-p-y/statically-typed-poco-queries#10]
This has to do with asherber/StaTypPocoQueries.PetaPoco#6 

```c#
public abstract class Parent
{
    public virtual int Id { get; set; }
}

[TableName("MyTable")]
public class Child : Parent
{
    [Column("Foo")]
    public override int Id { get; set; }
}

myDatabase.Query<Child>(c => c.Id == 1);
// expected: SELECT Foo FROM MyTable WHERE Foo = 1
// actual: SELECT Foo FROM MyTable WHERE Id = 1
```

Within my library, the `MemberInfo` that gets passed to my custom name extractor references `Parent` as the declaring type, so I don't see the `Column` property on `Child`, which is the actual runtime type. (Note that PetaPoco correctly sees `Child` as the type when building the SELECT and does use the intended column name.)

Is it possible for you to pass in a `MemberInfo` that relates to the actual runtime type, instead of the base type where the property is first declared?

**Comment by d-p-y** [comment]
Hi asherber,
Indeed, I confirm this behavior. In this case dotnet's  `Expression Tree` instance has `TypedParameterExpression` referencing `Child` but `MemberInfo` is taken from `Parent`. I'll look at it next days to see how to tackle it.

**Comment by asherber** [comment]
Thanks, @d-p-y. (Note: I edited my post so that the text correctly reflects the types in my example, but I see you understood anyway!)

**Comment by d-p-y** [comment]
@asherber I've addressed issue by extending `CustomNameExtractor` with extra parameter taken out of expression tree. New tests prove that it works. Can you check branch `issue10` and see if it works for you? Downside is that there's small API change.
Thanks

**Comment by asherber** [comment]
It works,  but it's awkward. In my name extractor, instead of figuring out the name from the `MemberInfo` passed in, now I'm doing `var realMemberInfo = type.GetMember(mi.Name)` -- grabbing the member on the real type with the same name as the original `MemberInfo` -- and then figuring out the column name from that. Isn't it possible for you to do that on your end, so that the `MemberInfo` you pass to the name extractor is correct? Then there wouldn't be an API change.

**Comment by d-p-y** [comment]
@asherber In your case 

```csharp
public abstract class Parent
{
    public virtual int Id { get; set; }
}

[TableName("MyTable")]
public class Child : Parent
{
    [Column("Foo")]
    public override int Id { get; set; }
}
```

field `Id` is really part of `Parent` definition. It is only actually implemented in `Child`. It seems awkward at first sight but it's logical for dotnet expression tree to return `Parent` in this case. 

Regarding `Isn't it possible for you to do that on your end` - I *could* but it seems to be questionable. Suppose that `PetaPoco` (or other `*Poco` library) user decides to put annotation in following way:

```csharp
public abstract class Parent2
{
    [Column("Foo")]
    public virtual int Id { get; set; }
}

public class Child2 : Parent2
{
    public override int Id { get; set; }
}
```
If we would implement logic that you ask (=for `Id` field return type `Child2` instead of `Parent2`) you wouldn't be able to find attribute. BTW this is one of the two new tests in this branch (`VirtualPropertiesIssue10UsesParentColumnAttribute`). It can be that such construct is illegal in `PetaPoco`. Still I'm not convinced that such strong opinion about "which field to return" is justified. That's why I believe place for
```csharp
(MemberInfo mi, Type t) => t.GetProperty(mi.Name)?.GetCustomAttribute<ColumnAttribute>()?.Name ?? mi.GetCustomAttribute<ColumnAttribute>()?.Name
```
belongs to your library.

Concerning API change - fair point. I've adapted code so that old methods in old F# module use old `customNameExtractor` definition (one without Type`). New methods that utilize Type parameter are in F# module `ExpressionToSqlV2`.

**Comment by asherber** [comment]
I see what you're saying about the "real" declaring type for the property, but that's not how reflection sees it. `typeof(Child).GetProperty("ID").DeclaringType` is `typeof(Child)` in both cases above.

In any case, the important thing is that when `db.Query<Child>(c => c.ID == 4)` produces `SELECT x FROM [Child] WHERE y = 4`, x and y should refer to the same column name. That is, the name that the *Poco libraries derive for the `SELECT` should be the same as the one that the binding libraries derive for the `WHERE`. I'm actually leveraging the mechanisms in PetaPoco (I think you're doing the same for the other libs), and the only way I can get this to work is if the `PropertyInfo` comes from `Child`, not `Parent`.

Here are 4 parent/child relationships. In all 4, because the `Column` attribute is inherited, I would expect the column name to be `Foo`. In both PetaPoco and NPoco, that's the result I get by using a mapper for the property on `Child`, so it makes sense to me that my custom name extractor would receive that `PropertyInfo`.

If you disagree, then I think I would prefer your original solution, with the breaking API change, to the new `ExpressionToSqlV2`. 

(Note that there is actually a bug in PetaPoco which I've discovered through testing this. In Example 2, PetaPoco returns `ID` for both the `SELECT` and the `WHERE`. I'll open an issue over there.)

```c#
// EXAMPLE 1
public class Parent
{
	public virtual int ID { get; set; }
}

public class Child : Parent
{
	[Column("Foo")]
	public override int ID { get; set; }
}

// EXAMPLE 2
public class Parent
{
	[Column("Foo")]
	public virtual int ID { get; set; }
}

public class Child : Parent
{
	public override int ID { get; set; }
}

// EXAMPLE 3
public class Parent
{
	[Column("Foo")]
	public virtual int ID { get; set; }
}

public class Child : Parent
{
	
}

// EXAMPLE 4
public class Parent
{
	[Column("NOTFoo")]
	public virtual int ID { get; set; }
}

public class Child : Parent
{
	[Column("Foo")]
	public override int ID { get; set; }
}
```

**Comment by d-p-y** [comment]
Concerning *"but that's not how reflection sees it. [...] in both cases above."* - I've checked and `DeclaringType` of `MemberInfo` seems to return type that was used as owner of property. Returns different values for parent and child:
![Screenshot_20220526_091526](https://user-images.githubusercontent.com/18470256/170442627-40b5de3e-3903-489b-b771-04cefa4aa225.png)
Expression tree *happens* to use abstract class with virtual property instead of subclass.

Anyway I've reverted "V2" changes. Now custom name mapper gets both vanilla `MemberInfo` and `Type` from expression tree (=body variable underneath).
![Screenshot_20220526_090645](https://user-images.githubusercontent.com/18470256/170443057-9d4fa66b-9c72-43ec-851a-1c52628ee7dc.png)

I agree that it is quirky that dotnet does it but I don't feel confident applying *"look for MemberInfo elsewhere"* lookup myself. If you are sure `PetaPoco` always expects such lookup then please apply it yourself (see my former comment for example code).

**Comment by asherber** [comment]
Thanks, that works for me. By the way, although I haven't dug into it, it looks like the same issue exists for the custom parameter value map -- the `PropertyInfo` I receive is not for the class in question, so I'm not able to see custom value converters on the class.

---

Just to continue the conversation, I would completely expect `GetProperty()` on Parent and Child to return different things -- when you're calling it on Parent, it doesn't know about Child.

Looking at the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.memberinfo.declaringtype?view=net-6.0), MS seems to mean `DeclaringType` to mean "the _closest_ place in the object hierarchy where the member is specified", not "the _earliest_ place in the object hierarchy". In my examples 1, 2, and 4 above, `typeof(Child).GetProperty("Id").DeclaringType` returns `Child` because I'm asking about `Child` and the property is specified there (even though it first occurs earlier in the hierarchy). For my example 3, the same expression would return `Parent`, because the Id property isn't specified in `Child`, just inherited. The docs happen to lay out examples very much like mine to show what happens.

In terms of this library, when I do `db.Query<Child>(c => c.ID == 4)`, the type represented by the lambda variable `c` is `Child`, so I would expect that the `MemberInfo` passed to my custom name extractor is from `typeof(Child).GetProperty("ID")` -- but that's not what I'm getting. With the second parameter you're providing, I can combine the `mi.Name` with the type to get the property info I'm really interested in.

FWIW, I'm actually letting PetaPoco do the actual name discovery. My old code was basically:

```c#
private string GetColumnName(MemberInfo mi)
{
	if (mi is PropertyInfo pi)
	{
		var mapper = Mappers.GetMapper(pi.DeclaringType, _db.DefaultMapper);
		var ci = mapper.GetColumnInfo(pi);
		return ci.ColumnName;
	}
}
```

I've now added a line `var realPI = type.GetProperty(pi.Name)`, which gets me the PI from the actual type in question, and then I pass `realPI` to the PP mapper functions.

**Comment by d-p-y** [comment]
Thanks, I've merged it into `master`. 
Thanks for spotting and reporting similar problem in `customParameterValueMap`, I've registered another issue for that. I'll release new version after that issue is addressed.

**Comment by d-p-y** [comment]
released as v1.16.0.0 @asherber

**Comment by asherber** [comment]
Yikes! Shouldn't this be v2, because of the breaking API change? Otherwise, updating this lib under one of the binding libs will fail to compile.

**Comment by d-p-y** [comment]
indeed it seems safer to try to communicate breaking changes by boosting version. I've unlisted v1.16.0.0 in nuget. Still won't help people with liberally specified version in their csproj. Myself, I prefer specifying exact version as it always produces same binaries. Will release it as v2 soon

**Comment by asherber** [comment]
Thanks, I've gotten so used to [semantic versioning](https://semver.org/) that I just take it for granted!

My binding library currently specifies a dependency on v1 of your library, so it explicitly won't work with v2 (but it would have happily tried to work with 1.16). I'll release a new version of my lib that specifies v2 of yours.

**Comment by d-p-y** [comment]
Thanks for the idea. I've re-released newest version to nuget as 2.0.0
